src/librustc_typeck/diagnostics.rs

```diff
@@ -4095,6 +4095,92 @@ assert_eq!(!Question::No, true);
 ```
 "##,
 
+E0604: r##"
+A cast to `char` was attempted on a type other than `u8`.
+
+Erroneous code example:
+
+```compile_fail,E0604
+0u32 as char; // error: only `u8` can be cast as `char`, not `u32`
+```
+
+As the error message indicates, only `u8` can be cast into `char`. Example:
+
+```
+let c = 86u8 as char; // ok!
+assert_eq!(c, 'V');
+```
+"##,
+
+E0605: r##"
+An invalid cast was attempted.
+
+Erroneous code examples:
+
+```compile_fail,E0605
+let x = 0u8;
+x as Vec<u8>; // error: non-primitive cast: `u8` as `std::vec::Vec<u8>`
+
+// Another example
+
+let v = 0 as *const u8; // So here, `v` is a `*const u8`.
+v as &u8; // error: non-primitive cast: `*const u8` as `&u8`
+```
+
+Only primitive types can be cast into each other. Examples:
+
+```
+let x = 0u8;
+x as u32; // ok!
+
+let v = 0 as *const u8;
+v as *const i8; // ok!
+```
+"##,
+
+E0606: r##"
+A cast between non-primitive types was attempted.
```

Comment:
I can totally appreciate getting it away from "non-scalar casts". That term is unlikely to make any sense when you come across it for the first time, especially as a beginner.

The full definition for what makes a valid `as` cast is [in the reference](https://doc.rust-lang.org/nightly/reference/expressions.html#type-cast-expressions). It can generally be reduced to the following categories:

* Explicitly casting something that's already a coercion (Deref coercion, trait object coercion, or references to raw pointers)
* Numeric casts (between two number types)
* Enum casts (from an enum with no fields in its variants to an integer)
* Pointer casts (between pointers, between thin pointers and integers, or from `&[T; N]` to `*const T`)
* Special primitive casts (from `u8` to `char`, or from `char` or `bool` to an integer)

I dunno how best to describe it without either explicitly listing all these options or papering over it with a short description and deliberately leaving some out. This also doesn't even count that not every casting error is an E0606, and that E0605 has very similar phrasing issues. It's easier to leave E0605 alone, though, since it's *actually* about casting with non-primitives. If we stuff an explanation of valid casts into this text, it might be worth duplicating it in E0605 since people may just run from one error to another before finally hitting this one.